### PR TITLE
Report kvikio I/O statistics per rank in the streaming engines

### PR DESCRIPTION
**Pull request by madsbk**
This PR enables KvikIO statistics on every rank and gathers them on the client. Depends on [rapidsai/kvikio#1036](https://github.com/rapidsai/kvikio/pull/1036), which adds the monitor that does the counting.

Each rank turns on counting when the engine is configured with `statistics=True`, and `StreamingEngine.gather_io_summary()` brings back one `kvikio.Summary` per rank, keyed by rank index.

```python
options = StreamingOptions(statistics=True)
with SPMDEngine(
    rapidsmpf_options=options.to_rapidsmpf_options(),
    executor_options=options.to_executor_options(),
    engine_options=options.to_engine_options(),
) as engine:
    pl.scan_parquet(path).select(pl.col("a").sum()).collect(engine=engine)
    for rank, summary in engine.gather_io_summary().items():
        print(f"--- rank {rank} ---")
        print(summary)
```

### What a rank reports

KvikIO renders the report, so this PR formats nothing. A single-rank run of a parquet scan:

```
KvikIO I/O summary
  wall time            122.55 ms
  busy time            18.40 ms (15.02 % of the wall time)
  busy bandwidth       66.44 MB/s
  operations           12 (12 read, 0 write)
  mean duration        3.90 ms
  bytes                1.17 MiB of 1.17 MiB requested (1.17 MiB read, 0 B written)
  errors               0
  backend POSIX        1.17 MiB in 12 ops, 46.83 ms, 26.11 MB/s
  backend GDS          unused
  backend MMAP         unused
  backend REMOTE_HTTP  unused
  backend REMOTE_HDFS  unused
```

**Busy time** counts only the stretches with a read in flight, so **busy bandwidth** measures the storage rather than the query: this scan spent 15 % of its wall time reading, and dividing by the whole span would have reported it at a tenth of the rate the disk was really giving.

### In the benchmarks

The PDS runners record the per-rank summaries on each iteration's record when `--rapidsmpf-statistics` is passed, so I/O stays queryable across a whole sweep rather than being printed once and lost.

`print_results_file.py` (new file) reads a results file back and prints it, since nothing existed that could. Timings and I/O side by side, one row per rank per iteration:

```
$ python -m cudf_polars.streaming.benchmarks.print_results_file pdsh-output.json

==============================================================================
run       : b6274527-a5d4-4791-9aa0-2a7a4ee3d307  (2026-08-20T08:35:11+00:00)
engine    : cudf-polars  frontend=ray
dataset   : /datasets/tpch-rs/scale-10-duckdb/  scale=10
workers   : 2  iterations=3
==============================================================================

Timings
   query  iters        min        max       mean
       1      3    0.1364s    0.3168s    0.2166s
       3      3    0.1616s    0.2154s    0.1824s
   total                                 0.3990s

I/O per rank
   query  iter  rank      ops         read       busy   busy%    bandwidth  backends
       1     0     0      796   312.09 MiB     64.0ms    9.8%     5.11GB/s  POSIX
       1     0     1      793   310.21 MiB     75.4ms   11.5%     4.31GB/s  POSIX
       1     1     0      788   306.54 MiB     40.1ms   29.1%     8.01GB/s  POSIX
       1     1     1      793   310.21 MiB     54.3ms   39.5%     5.99GB/s  POSIX
       1     2     0      788   306.54 MiB     37.3ms   18.9%     8.61GB/s  POSIX
       1     2     1      793   310.21 MiB     49.7ms   25.1%     6.54GB/s  POSIX
       3     0     0     1485   553.21 MiB     91.2ms   42.0%     6.36GB/s  POSIX
       3     0     1      793   339.17 MiB     82.1ms   37.8%     4.33GB/s  POSIX
       3     1     0     1474   549.74 MiB     68.6ms   40.0%     8.41GB/s  POSIX
       3     1     1      793   339.17 MiB     59.5ms   34.7%     5.98GB/s  POSIX
       3     2     0     1474   549.74 MiB     67.1ms   41.1%     8.59GB/s  POSIX
       3     2     1      793   339.17 MiB     61.3ms   37.6%     5.80GB/s  POSIX

  widest read skew: 1.63x (query 3, iteration 0)
```

### Caveats worth knowing

- **Counting is per process.** With Ray and Dask each rank has a process to itself, so a summary covers only cudf-polars. With SPMD cudf-polars shares your script's process, so KvikIO operations your own code performs are counted too.
- **Not all I/O is observed.** Per KvikIO's `Monitor` docs the cuFile async API on a working GDS system and the batch API report nothing, and anything cudf-polars reads outside KvikIO is invisible.
- **Timestamps are per rank.** Each monitor takes its own clock anchor, so across hosts the start and end times carry whatever NTP skew exists. Ratios within a rank, `busy_fraction` and the bandwidths, are unaffected.

### CODE REVIEW

File: docs/cudf/source/cudf_polars/profiling.md [hunk 1]
```diff
@@ -58,6 +58,87 @@ print(total)
 ```
 
 
+## I/O Statistics
+
+The same `statistics=True` also turns on [KvikIO I/O statistics][kvikio-stats] on every rank,
```
Comment (TomAugspurger) [inline docs/cudf/source/cudf_polars/profiling.md]: I don't think we should overload `statistics` here. Why not have a separate option for enabling kvikio / IO  monitoring?

If the goal is convenience, then we can provide a `StatisticsConfig(rapidsmpf=True, kvikio=True, ...)` object, and interpret `True` as setting each of those to true.

File: docs/cudf/source/cudf_polars/profiling.md [hunk 2]
```diff
@@ -58,6 +58,87 @@ print(total)
...
+  backend GDS          unused
+  backend MMAP         unused
+  backend REMOTE_HTTP  unused
+  backend REMOTE_HDFS  unused
+```
+
+Every row is also an attribute, `s.bytes_read`, `s.busy_ns` and so on. See the
+[KvikIO reference][kvikio-stats] for the full set.
+
+**Busy time** is the one worth explaining. It counts only the stretches with at least one read in
```
Comment (TomAugspurger) [inline docs/cudf/source/cudf_polars/profiling.md]: Is "Busy time" defined in kvikio or cudf-polars? If it's kvikio, we should document it there and link to it.

File: docs/cudf/source/cudf_polars/profiling.md [hunk 3]
```diff
@@ -58,6 +58,87 @@ print(total)
...
+Some I/O never reaches the monitor. On a system with working GDS the cuFile asynchronous API
+reports nothing, the batch API reports nothing, and anything cudf-polars reads outside KvikIO is
+invisible.
+
+[^shared-worker]: A Dask worker can host more than one rank if you run several engines, or other
+    Dask work, against one cluster. Neither is a recommended setup, and the summaries would be
+    mixed together.
+
+
+### In the benchmarks
```
Comment (TomAugspurger) [inline docs/cudf/source/cudf_polars/profiling.md]: I don't think we need this section in our user-facing docs.

File: python/cudf_polars/cudf_polars/streaming/benchmarks/print_results_file.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+        If the file contains no runs.
+    """
+    with path.open() as f:
+        runs = [json.loads(line) for line in f if line.strip()]
+    if not runs:
+        raise ValueError(f"{path} contains no runs")
+    return runs
+
+
+def iter_records(run: dict[str, Any]) -> Iterator[dict[str, Any]]:
```
Comment (TomAugspurger) [inline python/cudf_polars/cudf_polars/streaming/benchmarks/print_results_file.py]: benchmarks/utils.py has various dataclasses for these. I'd recommend using those instead of `dict[str, Any]`. Then you'll have type-cheked, safe access to fields like `run_id`, `frontend`, etc.

File: python/cudf_polars/tests/streaming/test_statistics.py [hunk 5]
```diff
@@ -58,3 +61,81 @@ def test_statistics(engine: StreamingEngine) -> None:
...
+
+    # Statistics are enabled on this fixture, so every rank is counting.
+    gathered = engine.gather_io_summary()
+    assert sorted(gathered) == list(range(engine.nranks))
+    summaries = list(gathered.values())
+    # Only some ranks may have been given a file to read, but not none of them.
+    assert sum(s.num_ops for s in summaries) > 0
+    assert sum(s.bytes_read for s in summaries) > 0
+
+    for s in summaries:
```
Comment (TomAugspurger) [inline python/cudf_polars/tests/streaming/test_statistics.py]: IIUC, the actually `Summary` object is defined in kvikio. If that's true, I'd prefer to minimize the number of assertions we make about it to just those things that cudf-polars controls/uses (which might be nothing, since I think we just return the summary unmodified).

File: python/cudf_polars/cudf_polars/engine/core.py [hunk 6]
```diff
@@ -15,6 +15,7 @@
 from typing import TYPE_CHECKING, Any, ClassVar, Self, TypeVar
 
 import cuda.core
+import kvikio
```
Comment (pentschev) [inline python/cudf_polars/cudf_polars/engine/core.py:18]: This now requires we add KvikIO’s Python package as a direct runtime dependency via `dependencies.yaml`.

File: python/cudf_polars/cudf_polars/streaming/benchmarks/print_results_file.py [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+    Returns
+    -------
+    The summaries, empty if the iteration recorded none.
+    """
+    raw = record.get("io_summaries")
+    if not isinstance(raw, dict):
+        # Absent, or written by a version that shaped it differently. The file
+        # comes from disk, so an unexpected shape is skipped rather than raised.
+        return {}
+    return {int(rank): s for rank, s in raw.items()}
```
Comment (pentschev) [inline python/cudf_polars/cudf_polars/streaming/benchmarks/print_results_file.py]: Could we validate each rank’s summary before returning it, and add a printer-level regression test? Although this helper says unexpected on-disk shapes are skipped, `{"0": null}` currently becomes `{0: None}` and `print_io_summaries()` then fails at `s.get(...)`. Skipping malformed entries (or reporting I/O as unavailable) would make the new results-file reader match its documented defensive behavior.

File: python/cudf_polars/tests/streaming/test_statistics.py [hunk 8]
```diff
@@ -58,3 +61,81 @@ def test_statistics(engine: StreamingEngine) -> None:
...
+    assert sum(s.num_ops for s in before) > 0
+
+    # Nothing between the clear and this gather, so the new span is empty.
+    after = engine.gather_io_summary()
+    assert len(after) == engine.nranks
+    assert sum(s.num_ops for s in after.values()) == 0
+    assert sum(s.bytes_transferred for s in after.values()) == 0
+
+
+def test_io_summary_is_off_without_statistics(
```
Comment (pentschev) [inline python/cudf_polars/tests/streaming/test_statistics.py]: Could this test reset the same engine from `statistics=False` to `statistics=True` rather than constructing a second engine? That would cover the newly added monitor stop/recreate lifecycle and verify the `rapidsmpf_options` update that benchmark collection relies on.

**Review by pentschev — CHANGES_REQUESTED** [review]

File: docs/cudf/source/cudf_polars/profiling.md (hunk 2 above)
Comment (madsbk) [reply docs/cudf/source/cudf_polars/profiling.md]: done

File: docs/cudf/source/cudf_polars/profiling.md (hunk 3 above)
Comment (madsbk) [reply docs/cudf/source/cudf_polars/profiling.md]: removed

File: python/cudf_polars/tests/streaming/test_statistics.py (hunk 5 above)
Comment (madsbk) [reply python/cudf_polars/tests/streaming/test_statistics.py]: good point, done

File: python/cudf_polars/cudf_polars/streaming/benchmarks/print_results_file.py (hunk 4 above)
Comment (madsbk) [reply python/cudf_polars/cudf_polars/streaming/benchmarks/print_results_file.py]: done

File: docs/cudf/source/cudf_polars/profiling.md (hunk 1 above)
Comment (madsbk) [reply docs/cudf/source/cudf_polars/profiling.md]: Good point, adding a dedicated `kvikio_statistics` option rather than overloading `statistics`. A separate knob is a better fit here, since we plan to add more expensive KvikIO statistics later, such as timeline and sampling monitors that can write trace files. Those should never be enabled implicitly by a RapidsMPF flag.

File: python/cudf_polars/cudf_polars/engine/core.py (hunk 6 above)
Comment (madsbk) [reply python/cudf_polars/cudf_polars/engine/core.py:18]: I don't think this is a new requirement. `py_run_cudf_polars` has `depends_on_kvikio` in `dependencies.yaml`, which adds `kvikio==26.10.*,>=0.0.0a0` to `python/cudf_polars/pyproject.toml`. This also predates this PR: `spmd.py` already imports `kvikio.defaults` on `main` to configure the thread pool.

File: python/cudf_polars/cudf_polars/streaming/benchmarks/print_results_file.py (hunk 7 above)
Comment (madsbk) [reply python/cudf_polars/cudf_polars/streaming/benchmarks/print_results_file.py]: I checked whether this is reachable, and it is not. All three engines omit non-counting ranks from `gather_io_summary()` rather than mapping them to `None`, so the writer’s `dataclasses.asdict(summary)` always produces a dict and cannot emit `{"0": null}`.

I have dropped the validation rather than adding to it. `io_summaries` is new in this PR, so no released version writes any other shape, and guarding against formats that never shipped would be dead weight from the moment this merges. I also removed the comment promising defensive skipping, so the helper no longer claims behavior it does not have.

File: python/cudf_polars/tests/streaming/test_statistics.py (hunk 8 above)
Comment (madsbk) [reply python/cudf_polars/tests/streaming/test_statistics.py]: That is already what happens. `streaming_engine_factory` resets the shared session-scoped engine in place, so the test already covers the monitor stop-and-recreate path. I added `assert reset is engine` to make that explicit.

The `rapidsmpf_options` part no longer applies; that gate was removed with the dedicated `kvikio_statistics` option.

File: python/cudf_polars/cudf_polars/engine/core.py (hunk 6 above)
Comment (pentschev) [reply python/cudf_polars/cudf_polars/engine/core.py:18]: Oh, my bad. I was looking at an older `main`. Sorry for the noise.

**Review by pentschev — APPROVED** [review]
LGTM, thanks Mads.

File: python/cudf_polars/cudf_polars/engine/dask.py [hunk 9]
```diff
@@ -1061,11 +1114,13 @@ def _reset(
             existing_kvikio_nthreads = existing_executor_options.get("kvikio_nthreads")
             if existing_kvikio_nthreads is not None:
                 executor_options.setdefault("kvikio_nthreads", existing_kvikio_nthreads)
+        executor_options.setdefault(
+            "kvikio_statistics", resolve_kvikio_statistics(executor_options)
+        )
```
[the remark this answers is not in the document]
Comment (madsbk) [reply python/cudf_polars/cudf_polars/engine/dask.py:1119]: Skipping this. Carrying it over would leak enabled statistics across resets: the test engine is session-scoped, and the baseline never sets `kvikio_statistics`, so `setdefault` would keep `True` for unrelated tests.

The right analogue is RAPIDSMPF statistics, which is also rebuilt from the options passed in. `quent_context` and `kvikio_nthreads` are carried for resource reasons, a live object and a disruptive pool rebuild. A boolean flag has neither.

No change made.

File: python/cudf_polars/cudf_polars/utils/config.py [hunk 10]
```diff
@@ -833,6 +844,9 @@ class StreamingExecutor:
     kvikio_nthreads: int = dataclasses.field(
         default_factory=lambda: resolve_kvikio_nthreads({})
     )
+    kvikio_statistics: bool = dataclasses.field(
+        default_factory=lambda: resolve_kvikio_statistics({})
+    )
```
Comment (TomAugspurger) [inline python/cudf_polars/cudf_polars/utils/config.py:850]: Can we delete `resolve_kvikio_statistics` and use `_bool_converter`?

```suggestion
    kvikio_statistics: bool = dataclasses.field(
        default_factory=_make_default_factory(
            f"{_env_prefix}__KVIKIO_STATISTICS", _bool_converter, default=False
        )
    )
 ```

Comment (madsbk) [reply python/cudf_polars/cudf_polars/utils/config.py:850]: Took the field half, that works as written. I couldn't delete `resolve_kvikio_statistics`, though. It's called from six sites across the three engines with an `executor_options` dict, while `_make_default_factory` only reads the environment. Same pattern as `resolve_kvikio_nthreads`. 

I did drop the hand-rolled bool parsing in favor of `_bool_converter`.

**Review by TomAugspurger — APPROVED** [review]
All good, thanks.

> It's called from six sites across the three engines with an executor_options dict, while _make_default_factory only reads the environment

That seems like a design issue :) We should have just one way of figuring out the value to use that works everywhere. https://github.com/NVIDIA/cudf/issues/23740 is tracking that general issue I think.
